### PR TITLE
feat: implement test enforcement and tag addition backend

### PR DESCRIPTION
**Pull request by virvirlopez**
# Description
Adding testing and tags Flow.
# How was the change tested
Ran it in local.

# Checklist

(Ideally, all boxes are checked by the time we merged the PR, if you don't know how to do any of these don't hesitate to say so in the PR and we'll help you out.)

- [X] I formatted my PR name according to [CONTRIBUTING.md](CONTRIBUTING.md)
- [ ] I added a news fragment to help populating the changelog as encouraged in [CONTRIBUTING.md](CONTRIBUTING.md)
- [ ] I added "Closes #<issue_number>" in the "Issue Information" section (if no issue, feel free to tick thick the box anyway).

### CODE REVIEW

File: dbt_sugar/core/task/base.py [hunk 1]
```diff
@@ -69,29 +69,96 @@ def get_not_documented_columns(
...
 
+    def __combine_two_list_without_duplicates(
+        self, list1: List[Any], list2: List[Any]
+    ) -> List[Any]:
+        """
+        Method to combine two list without duplicates.
+
+        :param list1: Optional[List[Any]]
+        :param list2: List[Any]
+        :return: List[Any]
```
Comment (bastienboutonnet) [inline dbt_sugar/core/task/base.py:80]: we use google docstrings in this project, you should be able to configure this in your IDE
the google dosctrings look like this:
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: dbt_sugar/core/task/doc.py [hunk 2]
```diff
@@ -17,6 +15,11 @@
 
 NUMBER_COLUMNS_TO_PRINT_PER_ITERACTION = 5
 
+DB_CONNECTORS = {
+    "postgres": PostgresConnector,
+    "snowflake": SnowflakeConnector,
+}
+
```
Comment (bastienboutonnet) [inline dbt_sugar/core/task/doc.py:22]: this is really nice!

File: dbt_sugar/core/task/doc.py [hunk 3]
```diff
@@ -49,27 +52,21 @@ def run(self) -> int:
...
-        elif type_of_connection == "snowflake":
-            columns_sql = SnowflakeConnector(
-                user=dbt_credentials.get("user", str()),
-                password=dbt_credentials.get("password", str()),
-                account=dbt_credentials.get("account", str()),
-                database=dbt_credentials.get("database", str()),
-            ).get_columns_from_table(model, schema)
+        connector = DB_CONNECTORS.get(type_of_connection)
+        if not connector:
+            print("The type of connector doesn't exists.")
```
Comment (bastienboutonnet) [inline dbt_sugar/core/task/doc.py:57]: could we raise instead of printing here?

File: dbt_sugar/core/task/doc.py [hunk 4]
```diff
@@ -26,7 +29,7 @@ class DocumentationTask(BaseTask):
 
     def __init__(self, flags: FlagParser, dbt_profile: DbtProfile) -> None:
         super().__init__()
-        self.columns_to_update: Dict[str, str] = {}
+        self.columns_to_update: Dict[str, Dict[str, Any]] = {}
```
Comment (bastienboutonnet) [inline dbt_sugar/core/task/doc.py:32]: I'm a little confused where this gets filled in

File: tests/postgres_test.py [hunk 5]
```diff
@@ -6,7 +6,7 @@
 def test_generate_connection():
     from dbt_sugar.core.connectors.postgres_connector import PostgresConnector
 
-    conn = PostgresConnector(**CREDENTIALS).generate_connection()
+    conn = PostgresConnector(**CREDENTIALS)
```
Comment (bastienboutonnet) [inline tests/postgres_test.py:9]: we make is so that when we init we always create a connection?

**Review by bastienboutonnet — COMMENTED** [review]
I left a few questions because I don't follow everything.

Also, our test coverage is taking a serious hit and from what I can see there are things we can easily add tests for. I really like the idea of testing as much as possible and some of these unhit lines seem to be quite important.

File: dbt_sugar/core/task/doc.py (hunk 3 above)
Comment (virvirlopez) [reply dbt_sugar/core/task/doc.py:57]: yes, that will be better

File: dbt_sugar/core/task/doc.py (hunk 4 above)
Comment (virvirlopez) [reply dbt_sugar/core/task/doc.py:32]: This is the response of the UI
{"COLUMN_NAME: {"description": "x", "tags" [], "tests": []}}

**Review by bastienboutonnet — COMMENTED** [review]
This is looking pretty good I have a few suggestion to improve the wording of some of the docstrings and some questions/things to consider.

File: tests/postgres_test.py (hunk 5 above)
Comment (virvirlopez) [reply tests/postgres_test.py:14]: yes, well is not a connection is the engine for making later the connection.

Comment (bastienboutonnet) [reply tests/postgres_test.py:14]: Got it! Sounds good anyway, just wanted to make sure.

File: tests/postgres_test.py [hunk 6]
```diff
@@ -19,3 +24,37 @@ def test_get_columns_from_table():
...
+        ),
+    ],
+)
+def test_run_test(mocker, test_name, schema, table, column_name, result):
+    run_query_test = mocker.patch(
+        "dbt_sugar.core.connectors.postgres_connector.PostgresConnector.run_query_test"
+    )
+    postgres_connector = PostgresConnector(**CREDENTIALS)
+    postgres_connector.run_test(test_name, schema, table, column_name)
+    run_query_test.assert_has_calls(result)
```
Comment (bastienboutonnet) [inline tests/postgres_test.py:60]: Actually we don't (and in fact shouldn't) be mocking tests that we do in postgress because we can actually test it fully end to end against the database as it gets set up in docker even on the github actions.

Don't you think it would be a good idea?

File: dbt_sugar/core/connectors/base.py [hunk 7]
```diff
@@ -17,34 +18,92 @@ class BaseConnector(ABC):
...
+        Returns:
+            boolean: True if the test is okay, and False if the test doesn't pass.
+        """
+        TESTS = {
+            "unique": f"""select count(*) as errors from(
+                select {column} from {schema}.{table} where {column} is not null group by {column} having count(*) > 1 )
+                errors""",
+            "not_null": f"select count(*) as errors from {schema}.{table} where {column} is null",
+        }
+        query = TESTS[test_name]
```
[the remark this answers is not in the document]
Comment (virvirlopez) [reply dbt_sugar/core/connectors/base.py:88]: I don't think is necessary because we give the options to the user, so that raise will never happens.

File: dbt_sugar/core/connectors/postgres_connector.py [hunk 8]
```diff
@@ -17,15 +17,23 @@ class PostgresConnector(BaseConnector):
...
     """
 
-    def __init__(self, user: str, password: str, database: str, host: str = "localhost") -> None:
+    def __init__(
+        self,
+        user: str,
+        password: str,
+        database: str,
+        host: str = "localhost",
+        account: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (virvirlopez) [reply dbt_sugar/core/connectors/postgres_connector.py:26]: I want to have the same parameters for all the databases so we can call it with the same parameters, we can do the clean way of having the dict with the options, and instantiated whatever object.

File: tests/postgres_test.py [hunk 9]
```diff
@@ -1,12 +1,21 @@
...
+import pytest
 import sqlalchemy
 
-CREDENTIALS = dict(user="dbt_sugar_test_user", password="magical_password", database="dbt_sugar")
+from dbt_sugar.core.connectors.postgres_connector import PostgresConnector
+
+CREDENTIALS = {
+    "username": "dbt_sugar_test_user",
+    "password": "magical_password",
+    "database": "dbt_sugar",
```
Comment (bastienboutonnet) [inline tests/postgres_test.py:11]: Ok I think I know what is making the test fail, you're not providing some of the required info for a postgress db.
We need a host and a port number. In theory the credentials validation would complain but since you're feeding it directly into the connector rather than via the dbt profile parsing we kind of skip this in the test.

I had a similar problem before.

File: dbt_sugar/core/task/doc.py [hunk 10]
```diff
@@ -47,29 +50,15 @@ def run(self) -> int:
...
-                    f"Could not connect to your postgres database, check your profiles.yml\n{e}"
-                )
-        elif type_of_connection == "snowflake":
-            columns_sql = SnowflakeConnector(
-                user=dbt_credentials.get("user", str()),
-                password=dbt_credentials.get("password", str()),
-                account=dbt_credentials.get("account", str()),
-                database=dbt_credentials.get("database", str()),
-            ).get_columns_from_table(model, schema)
+        connector = DB_CONNECTORS.get(dbt_credentials.get("type", ""))
```
Comment (bastienboutonnet) [inline dbt_sugar/core/task/doc.py:53]: can we name this `connector_type` so that we don't think that it is a connector object by mistake?

File: dbt_sugar/core/task/doc.py [hunk 11]
```diff
@@ -158,9 +149,37 @@ def orchestrate_model_documentation(self, model_name: str, columns_sql: List[str
         not_documented_columns = self.get_not_documented_columns(content, model_name)
         self.document_columns(not_documented_columns)
 
-        self.update_column_descriptions(self.columns_to_update)
+        # Will remove from the update the test that doesn't run correctly.
```
Comment (bastienboutonnet) [inline dbt_sugar/core/task/doc.py:152]: I don't understand that comment can you rephrase it?

File: dbt_sugar/core/task/doc.py [hunk 12]
```diff
@@ -158,9 +149,37 @@ def orchestrate_model_documentation(self, model_name: str, columns_sql: List[str
...
+        self.check_tests(schema, model_name)
+        # Method to update the description test and tags from the model.
```
Comment (bastienboutonnet) [inline dbt_sugar/core/task/doc.py:154]: not sure if we need that comment here I think the function name is pretty clear.

File: dbt_sugar/core/task/doc.py [hunk 13]
```diff
@@ -158,9 +149,37 @@ def orchestrate_model_documentation(self, model_name: str, columns_sql: List[str
...
+        self.update_model_description_test_tags(path, model_name, self.column_update_payload)
+        # Method to update the descriptions in all the schemas.yml
```
Comment (bastienboutonnet) [inline dbt_sugar/core/task/doc.py:156]: same here.

File: dbt_sugar/core/task/doc.py [hunk 14]
```diff
@@ -158,9 +149,37 @@ def orchestrate_model_documentation(self, model_name: str, columns_sql: List[str
...
+        self.update_column_descriptions(self.column_update_payload)
+
         return 0
 
+    def check_tests(self, schema: str, model_name: str) -> None:
+        """
+        Method to run and add test into a schema.yml, this method will:
+
+        Run the tests and if they have been successful it will add them into the schema.yml.
```
Comment (bastienboutonnet) [inline dbt_sugar/core/task/doc.py:165]:
```suggestion
        - Run the tests against the database
        - check whether they are passing 
        - if so, it will add them to the schema.yml.
```

File: dbt_sugar/core/connectors/base.py [hunk 15]
```diff
@@ -17,34 +18,84 @@ class BaseConnector(ABC):
...
 
-    @abstractmethod
-    def generate_connection(self) -> sqlalchemy.engine:
+    def __init__(
+        self,
+        connection_params: Dict[str, str],
+    ) -> None:
         """
-        Method that creates the connection.
+        Init method to instanciate the credentials.
```
Comment (bastienboutonnet) [inline dbt_sugar/core/connectors/base.py:26]: I think that's kind of not really what this does.

File: dbt_sugar/core/connectors/base.py [hunk 16]
```diff
@@ -17,34 +18,84 @@ class BaseConnector(ABC):
...
 
-        Returns:
-            sqlalchemy.engine: Engine to connect to the database.
+        Args:
+            connection_params (Dict[str, str]): Dict with connection parameters.
```
Comment (bastienboutonnet) [inline dbt_sugar/core/connectors/base.py:29]:
```suggestion
            connection_params (Dict[str, str]): Dict containing database connection parameters and credentials. 
```

File: dbt_sugar/core/connectors/base.py [hunk 17]
```diff
@@ -17,34 +18,84 @@ class BaseConnector(ABC):
...
         """
-        logging.info("Genetating connection")
+        self.connection_url = sqlalchemy.engine.url.URL(
+            drivername="postgresql+psycopg2",
```
Comment (bastienboutonnet) [inline dbt_sugar/core/connectors/base.py:32]: I don't understand why be base is using a postrgres connector? To me the base shouldn't implement any of the connection URL stuff that's a responsibility of the concrete classes

File: dbt_sugar/core/connectors/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (bastienboutonnet) [reply dbt_sugar/core/connectors/base.py:77]: Oh yeah that makes sense

File: dbt_sugar/core/connectors/postgres_connector.py [hunk 18]
```diff
@@ -17,47 +17,22 @@ class PostgresConnector(BaseConnector):
...
     """
 
-    def __init__(self, user: str, password: str, database: str, host: str = "localhost") -> None:
+    def __init__(
+        self,
+        connection_params: Dict[str, str],
+    ) -> None:
         """
-        Init method to instanciatee the credentials.
+        Init method to instanciate the credentials.
```
Comment (bastienboutonnet) [inline dbt_sugar/core/connectors/postgres_connector.py:25]: again I think that's slightly innacurate

**Review by bastienboutonnet — COMMENTED** [review]
I have a few more comments with some docstring suggestions and a question about something that puzzles me in the `BaseConnector` but other than that I think we should be good to merge soon.

Also, I feel we could be a bit more aggressive of unit test coverage, I see quite a significant amount of unhit lines, dunno what you think.

File: tests/postgres_test.py [hunk 19]
```diff
@@ -15,7 +26,41 @@ def test_get_columns_from_table():
...
+        ),
+    ],
+)
+def test_run_test(mocker, test_name, schema, table, column_name, result):
+    execute_and_check = mocker.patch(
+        "dbt_sugar.core.connectors.postgres_connector.PostgresConnector.execute_and_check"
+    )
+    postgres_connector = PostgresConnector(CREDENTIALS)
+    postgres_connector.run_test(test_name, schema, table, column_name)
+    execute_and_check.assert_has_calls(result)
```
Comment (bastienboutonnet) [inline tests/postgres_test.py:66]: Can we not do an actual test? We do spring up a postgres docker so I think we can go all the way here don't you think?

File: tests/postgres_test.py [hunk 20]
```diff
@@ -15,7 +26,41 @@ def test_get_columns_from_table():
...
+            "not_null",
+            "schema",
+            "table",
+            "column",
+            [call("select count(*) as errors from schema.table where column is null")],
+        ),
+    ],
+)
+def test_run_test(mocker, test_name, schema, table, column_name, result):
+    execute_and_check = mocker.patch(
```
Comment (bastienboutonnet) [inline tests/postgres_test.py:61]: I don't think we need to mock. We have a postgres instance.

**Review by bastienboutonnet — COMMENTED** [review]
Also some question around mocking postgres connector which I don't think we should do since we have the luxury of actually going all the way in our testing.

File: dbt_sugar/core/task/doc.py (hunk 10 above)
Comment (virvirlopez) [reply dbt_sugar/core/task/doc.py:53]: sure!

Comment (virvirlopez) [reply dbt_sugar/core/task/doc.py:53]: I think that is better the name connector,
because we are creating a connector in that variable, not a type.

**Review by bastienboutonnet — APPROVED** [review]
Ok this is now really good now thanks for taking the time to add the tests against the db.

BTW with this PR we have out MVP!
